### PR TITLE
SONAR-6259 Move components persistence from batch to Compute Engine

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+  private static boolean isProjectOrModule(BatchReport.Component reportComponent) {
+    return reportComponent.getType().equals(Constants.ComponentType.PROJECT) || reportComponent.getType().equals(Constants.ComponentType.MODULE);
+  }
+
+  private static String getScope(BatchReport.Component reportComponent) {
+    switch (reportComponent.getType()) {
+      case PROJECT:
+        return Scopes.PROJECT;
+      case MODULE:
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java]: New MAJOR Either merge this case with the identical one on line "182" or change one of the implementations. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1871)

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+      case DIRECTORY:
+        return Scopes.DIRECTORY;
+      case FILE:
+        return Scopes.FILE;
+    }
+    throw new IllegalArgumentException(String.format("Unknown type '%s'", reportComponent.getType()));
+  }
+
+  private static String getQualifier(BatchReport.Component reportComponent) {
+    switch (reportComponent.getType()) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java]: New MAJOR Add a default case to this switch. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        return Scopes.PROJECT;
+      case DIRECTORY:
+        return Scopes.DIRECTORY;
+      case FILE:
+        return Scopes.FILE;
+    }
+    throw new IllegalArgumentException(String.format("Unknown type '%s'", reportComponent.getType()));
+  }
+
+  private static String getQualifier(BatchReport.Component reportComponent) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java]: New MAJOR The Cyclomatic Complexity of this method "getQualifier" is 11 which is greater than 10 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+        existingComponent.setDescription(newComponent.description());
+        isUpdated = true;
+      }
+    }
+
+    if (!StringUtils.equals(existingComponent.moduleUuid(), newComponent.moduleUuid())) {
+      existingComponent.setModuleUuid(newComponent.moduleUuid());
+      isUpdated = true;
+    }
+    if (!existingComponent.moduleUuidPath().equals(newComponent.moduleUuidPath())) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java]: New BLOCKER NullPointerException might be thrown as 'moduleUuidPath' is nullable here [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2259)
New CRITICAL Possible null pointer dereference in org.sonar.server.computation.step.PersistComponentsStep.updateComponent(ComponentDto, ComponentDto) due to return value of called method [rule](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+    return isUpdated;
+  }
+
+  private static boolean isProjectOrModule(BatchReport.Component reportComponent) {
+    return reportComponent.getType().equals(Constants.ComponentType.PROJECT) || reportComponent.getType().equals(Constants.ComponentType.MODULE);
+  }
+
+  private static String getScope(BatchReport.Component reportComponent) {
+    switch (reportComponent.getType()) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java]: New MAJOR Add a default case to this switch. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/FeedComponentsCacheStep.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    ComponentDto componentDto = componentDtosByKey.get(componentKey);
+    if (componentDto == null) {
+      computeComponentsRefCache.addComponent(componentRef, new ComputeComponentsRefCache.ComputeComponent(componentKey, Uuids.create()));
+    } else {
+      computeComponentsRefCache.addComponent(componentRef, new ComputeComponentsRefCache.ComputeComponent(componentKey, componentDto.uuid()));
+    }
+
+    for (Integer childRef : reportComponent.getChildRefList()) {
+      // If current component is not a module or a project, we need to keep the parent reference to the nearest module
+      BatchReport.Component nextModuleParent = !reportComponent.getType().equals(Constants.ComponentType.PROJECT) && !reportComponent.getType().equals(Constants.ComponentType.MODULE) ?
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/FeedComponentsCacheStep.java]: New MINOR Split this 184 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistFileSourcesStep.java [hunk 7]
```diff
@@ -131,12 +134,13 @@ private void persistSource(FileSourcesContext fileSourcesContext, ComputeFileSou
     String dataHash = DigestUtils.md5Hex(data);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistFileSourcesStep.java]: MAJOR Use a stronger encryption algorithm than MD5. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2070)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 new issues:
* 1 blocker
* 1 critical
* 4 major
* 1 minor

Details:
* [Split this 184 characters long line (which is greater than 180 authorized).](https://github.com/SonarSource/sonarqube/blob/6d021e60119379ecf1e2c96f3aff02e4955c0c11/server/sonar-server/src/main/java/org/sonar/server/computation/step/FeedComponentsCacheStep.java#L90) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)
* [Add a default case to this switch.](https://github.com/SonarSource/sonarqube/blob/6d021e60119379ecf1e2c96f3aff02e4955c0c11/server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java#L195) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)
* [Add a default case to this switch.](https://github.com/SonarSource/sonarqube/blob/6d021e60119379ecf1e2c96f3aff02e4955c0c11/server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java#L181) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)
* [NullPointerException might be thrown as 'moduleUuidPath' is nullable here](https://github.com/SonarSource/sonarqube/blob/6d021e60119379ecf1e2c96f3aff02e4955c0c11/server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java#L164) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2259)
* [The Cyclomatic Complexity of this method "getQualifier" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonarqube/blob/6d021e60119379ecf1e2c96f3aff02e4955c0c11/server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java#L194) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
* [Either merge this case with the identical one on line "182" or change one of the implementations.](https://github.com/SonarSource/sonarqube/blob/6d021e60119379ecf1e2c96f3aff02e4955c0c11/server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java#L184) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1871)
* [Possible null pointer dereference in org.sonar.server.computation.step.PersistComponentsStep.updateComponent(ComponentDto, ComponentDto) due to return value of called method](https://github.com/SonarSource/sonarqube/blob/6d021e60119379ecf1e2c96f3aff02e4955c0c11/server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java#L164) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)

File: sonar-batch-protocol/src/main/protobuf/batch_report.proto [hunk 8]
```diff
@@ -83,9 +83,7 @@ message Component {
   optional string description = 16;
 
   // temporary fields during development of computation stack
-  optional int64 id = 13;
```
Comment (simonbrandhof) [inline sonar-batch-protocol/src/main/protobuf/batch_report.proto]: as long as this protobuf file does not support past versions, field ids can be compressed (no holes in sequence of numbers)

**Comment by simonbrandhof** [comment]
We know that CE design is being changed, so LGTM for now.

File: sonar-batch-protocol/src/main/protobuf/batch_report.proto (hunk 8 above)
Comment (julienlancelot) [inline sonar-batch-protocol/src/main/protobuf/batch_report.proto]: Good point !

**Pull request merged**